src/ralph/backends/database/clickhouse.py

```diff
@@ -0,0 +1,358 @@
+
+            try:
+                event_id, emission_time = ClickHouseDatabase._validate_and_parse(
+                    statement
+                )
+
+                document = (event_id, emission_time, statement, json.dumps(statement))
+
+                yield document
+            except BadFormatException as exc:
+                if ignore_errors:
+                    logger.warning(str(exc))
+                    continue
+                raise
+
+    def bulk_import(self, batch: List, ignore_errors: bool = False) -> int:
+        """Inserts a batch of documents into the selected database table."""
+        try:
+            # ClickHouse does not do unique keys. This is a "best effort" to
+            # at least check for duplicates in each batch. Overall ID checking
+            # against the database happens upstream in the POST / PUT methods.
+            #
+            # As opposed to Mongo, the entire batch is guaranteed to fail here
+            # if any dupes are found.
+            found_ids = {x[0] for x in batch}
+
+            if len(found_ids) != len(batch):
+                raise BackendException("Duplicate IDs found in batch")
+
+            self.client.insert(
+                self.event_table_name,
+                batch,
+                column_names=[
+                    "event_id",
+                    "emission_time",
+                    "event",
+                    "event_str",
+                ],
+                # Allow ClickHouse to buffer the insert, and wait for the
+                # buffer to flush. Should be configurable, but I think these are
+                # reasonable defaults.
+                settings={"async_insert": 1, "wait_for_async_insert": 1},
+            )
+        except (ClickHouseError, BackendException) as error:
+            if not ignore_errors:
+                raise BackendException(*error.args) from error
+            logger.warning(
+                "Bulk import failed for current chunk but you choose to ignore it.",
+            )
+            # There is no current way of knowing how many rows from the batch
+            # succeeded, we assume 0 here.
+            return 0
+
+        logger.debug("Inserted %s documents chunk with success", len(batch))
+
+        return len(batch)
+
+    def put(
+        self,
+        stream: Union[TextIO, List],
+        chunk_size: int = 500,
+        ignore_errors: bool = False,
+    ) -> int:
+        """Writes documents from the `stream` to the instance table."""
+        logger.debug(
+            "Start writing to the %s table of the %s database (chunk size: %d)",
+            self.event_table_name,
+            self.database,
+            chunk_size,
+        )
+
+        rows_inserted = 0
+        batch = []
+        for document in self.to_documents(stream, ignore_errors=ignore_errors):
+            batch.append(document)
+            if len(batch) < chunk_size:
```

Comment:
Following my previous remark about `batch` type, consuming a generator would be more efficient than a list. Hoping the ClickHouse client supports it. :crossed_fingers:

Reply:
Sadly ClickHouse doesn't seem to support accepting insert statements from generators, so I've left this one the same.